python/tank/util/shotgun.py

```diff
@@ -312,25 +285,80 @@ def create_sg_connection(user="default"):
+              is an sg entity dictionary (keys type/id) corresponding to to the user used
+              to connect to the app store.
     """
     global g_app_store_connection
     
-    if g_app_store_connection is None:
-        g_app_store_connection = __create_sg_connection(__get_app_store_config(), evaluate_script_user=True)
+    # maintain a cache for performance
+    if g_app_store_connection:
+        return g_app_store_connection
+    
+    # get app store credentials from shotgun
+    client_site_sg = get_sg_connection()
+    (script_name, script_key) = __get_app_store_key_from_shotgun(client_site_sg)
+    
+    # connect to the app store
+    config_data = {}
+    config_data["host"] = client_site_sg.base_url
+    config_data["api_script"] = script_name
+    config_data["api_key"] = script_key
+    config_data["http_proxy"] = client_site_sg.config.raw_http_proxy
+    app_store_sg = __create_sg_connection(config_data)
     
-    return g_app_store_connection
+    # now figure out the entity id for the script user    
+    script_user = app_store_sg.find_one("ApiUser", [["firstname", "is", script_name]], fields=["type", "id"])
+    if script_user is None:
+        raise TankError("Could not evaluate the current App Store User! Please contact support.")
+    
+    return (app_store_sg, script_user)
+
+
+def __get_app_store_key_from_shotgun(sg_connection):
+    """
+    Given a Shotgun url and script credentials, fetch the app store key
+    for this shotgun instance using a special controller method.
+    Returns a tuple with (app_store_script_name, app_store_auth_key)
+    
+    :param sg_connection: SG connection to the client site for which
+                          app store credentials should be retrieved.
+    :returns: tuple of strings with contents (script_name, script_key)
+    """
+    # handle proxy setup by pulling the proxy details from the main shotgun connection
+    if sg_connection.config.proxy_server:
+        
+        if sg_connection.config.proxy_user and sg_connection.config.proxy_pass:
+            auth_string = "%s:%s@" % (sg_connection.config.proxy_user, sg_connection.config.proxy_pass)
+        else:
+            auth_string = ""
+        proxy_addr = "http://%s%s:%d" % (auth_string, 
+                                         sg_connection.config.proxy_server, 
+                                         sg_connection.config.proxy_port)
+        proxy_support = urllib2.ProxyHandler({sg_connection.config.scheme : proxy_addr})
+        opener = urllib2.build_opener(proxy_support)
+        urllib2.install_opener(opener)
+    
+    # now connect to our site and use a special url to retrieve the app store script key
+    session_token = sg_connection.generate_session_token()
+    post_data = {"session_token": session_token }
+    response = urllib2.urlopen("%s/api3/sgtk_install_script" % sg_connection.base_url, urllib.urlencode(post_data) ) 
```

Comment:
nitpicking, but this isn't consistent with your coding style it seems, you have an extra whitespace before the )